### PR TITLE
[wip] Fix public protobuf interface - wip

### PR DESCRIPTION
**Pull request by Yangqing**
This is an ongoing fix to the protobuf issue, mainly to address 2 things:

(1) we have random protobuf fixes trying to patch an already flaky system (for example, dual install from anaconda and brew). This diff aims to basically use standard packages as much as possible (protobuf cmake config files, or FindProtoBuf.cmake) and then enforce the build script to explicitly set paths.

(2) We need protobuf to be in the public interface of Caffe2. This PR adds it.

(3) We will most likely need a protobuf diagnostic tool / script. TBD.

Firing a PR so that we can launch build tests.

### CODE REVIEW

File: cmake/public/protobuf.cmake
Comment (pjh5) [inline cmake/public/protobuf.cmake]: Will this question always be true?

File: cmake/public/protobuf.cmake
Comment (pjh5) [inline cmake/public/protobuf.cmake]: nit: protobuf::libprotobuf

**Comment by Yangqing** [comment]
Discovered issue in centos build that is being fixed at #1965.

File: cmake/public/protobuf.cmake
Comment (Yangqing) [reply cmake/public/protobuf.cmake]: nice catch :)

File: cmake/public/protobuf.cmake
Comment (Yangqing) [reply cmake/public/protobuf.cmake]: I think this should always be true. Let me just remove it, indeed.

**Comment by Yangqing** [comment]
@caffe2bot test this please.

**Comment by Yangqing** [comment]
@caffe2bot  retest this please. Seems that jenkins had a hiccup.

**Comment by Yangqing** [comment]
@caffe2bot retest this please. Jenkins docker update.

**Comment by Yangqing** [comment]
@pjh5 @orionr PTAL? :)

Cc @prigoyal FYI

**Review by pjh5 — APPROVED** [review]
Saw a few comment / style nits but I'm looking forward to getting this merged sooner rather than later

File: cmake/ProtoBuf.cmake
Comment (orionr) [inline cmake/ProtoBuf.cmake]: Nit: prooerly -> properly

**Review by orionr — APPROVED** [review]
Yes please!

**Comment by orionr** [comment]
@caffe2bot retest this please

File: cmake/public/protobuf.cmake
Comment (pietern) [inline cmake/public/protobuf.cmake]: This says glog but should be protobuf

File: .jenkins/build.sh
Comment (pietern) [inline .jenkins/build.sh]: Is this still debugging stuff? Extra newlines can go.

File: cmake/Caffe2Config.cmake.in
Comment (pietern) [inline cmake/Caffe2Config.cmake.in]: glog -> protobuf

File: cmake/ProtoBuf.cmake
Comment (pietern) [inline cmake/ProtoBuf.cmake]: Should go before merging

File: cmake/public/protobuf.cmake
Comment (pietern) [inline cmake/public/protobuf.cmake]: Trailing whitespace

**Review by pietern — COMMENTED** [review]
Nice work! Mostly typo's in this review.

**Comment by Yangqing** [comment]
Thanks folks - will merge when the test passes.

**Comment by Yangqing** [comment]
Confirmed that test failure is due to docker image building; new docker images are being built here: https://ci.pytorch.org/jenkins/job/caffe2-docker-trigger/65/. Landing.

**Pull request merged**
